prow/git/git_command_test.go

```diff
@@ -0,0 +1,273 @@
+			},
+			expected: Repo{
+				Dir:  tempdir.recorded(5),
+				git:  gitbin,
+				base: urlMustParse("https://github.com"),
+				repo: "org/repo",
+				user: "user",
+				pass: "password",
+			},
+		},
+		{
+			name:                 "error path with remote url and user, retries exceeded",
+			base:                 "https://github.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 4,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+			},
+			err: true,
+		},
+		{
+			name:                 "happy path with GHE",
+			base:                 "https://corp.github.initech.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 0,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@corp.github.initech.com/org/repo", tempdir.recorded(8) + "/org/repo.git"},
+				{"", gitbin, "clone", tempdir.recorded(8) + "/org/repo.git", tempdir.recorded(9)},
+			},
+			expected: Repo{
+				Dir:  tempdir.recorded(9),
+				git:  gitbin,
+				base: urlMustParse("https://corp.github.initech.com"),
+				repo: "org/repo",
+				user: "user",
+				pass: "password",
+			},
+		},
+	}
+	defer func() { execWithDir = defaultExecWithDir }()
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			simulatedFailures := 0
+			var actualCommands [][]string
+			execWithDir = func(dir, name string, args ...string) ([]byte, error) {
+				actualCommands = append(actualCommands, append([]string{dir, name}, args...))
+				if simulatedFailures < tc.simulateExecFailures {
+					simulatedFailures++
+					return nil, fmt.Errorf("Simulated failure %d", simulatedFailures)
+				}
+				return nil, nil
+			}
+			baseURL, err := url.Parse(tc.base)
+			if err != nil {
+				t.Errorf("Unexpected error parsing url %s: %v", tc.base, err)
+			}
+			client, err := NewClient(baseURL)
+			if err != nil {
+				t.Errorf("Unexpected error creating Client: %v", err)
+			}
+			client.SetCredentials(tc.user, tc.tokenGenerator)
+
+			actual, err := client.Clone(tc.repo)
+			switch {
+			case !tc.err && err != nil:
+				t.Errorf("Unexpected error result: %v", err)
+			case tc.err && err == nil:
+				t.Errorf("Expected error, got nil")
+			}
+			if tc.err && actual != nil {
+				t.Errorf("Unexpected non-nil result")
+			}
+			if !tc.err {
```

Comment:
Don't check this conditionally -- if you expect an error, just expect a `nil` Repo